### PR TITLE
feat: s2s auto config schema

### PR DESCRIPTION
**Pull request by elboy3**
https://app.shortcut.com/galileo/story/8103/dq-auto-improvements-config-schema

As suggested by Bogdan and Jon:
- https://github.com/rungalileo/dataquality/pull/767#discussion_r1347950908
- https://github.com/rungalileo/dataquality/pull/767/files#r1347886773

### CODE REVIEW

File: dataquality/integrations/seq2seq/auto.py [hunk 1]
```diff
@@ -84,47 +81,25 @@ def auto(
     One of `hf_data`, `train_data` should be provided. If neither of those are, a
```
Comment (dcaustin33) [inline dataquality/integrations/seq2seq/auto.py:81]: Update these?

File: dataquality/integrations/seq2seq/s2s_trainer.py [hunk 2]
```diff
@@ -18,26 +19,18 @@
...
-# Check more out here: https://huggingface.co/docs/transformers/v4.30.0/en/main_classes/text_generation#transformers.GenerationConfig
-MAX_NEW_TOKENS = 16
-TEMPERATURE = 0.2  # Keep this low for now
-TOP_P = 1
-TOP_K = 50
 # Generate params
 GENERATE_ON_TRAIN = False
 # Tokenization params
 MAX_INPUT_LENGTH = 512
 MAX_TARGET_LENGTH = 128
```
Comment (dcaustin33) [inline dataquality/integrations/seq2seq/s2s_trainer.py:32]: Dont need these

**Review by dcaustin33 — COMMENTED** [review]
Looks real good to me besides cleaning up the docstrings and other stuff!

File: dataquality/integrations/seq2seq/schema.py [hunk 3]
```diff
@@ -1 +1,94 @@
...
+        and `test_data`. If provided, train_data, val_data, and test_data are ignored
+    :param train_path: Optional path to training data file to use. Can be one of
+        * Path to a local file
+        * Huggingface dataset hub path
+    :param val_path: Optional path to validation data to use. Can be one of
+        * Path to a local file
+        * Huggingface dataset hub path
+    :param test_path: Optional test data to use. Can be one of
+        * Path to a local file
+        * Huggingface dataset hub path
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/schema.py:27]: I see that `train_path`, `val_path` and `test_path` are not used yet. Is that for later ? By HF path do you mean a path to the parquet file ? One thing with that is that it might require to also pass the split (or we could assume it's one of train/training).

File: dataquality/integrations/seq2seq/schema.py [hunk 4]
```diff
@@ -1 +1,94 @@
...
+    :param max_target_tokens: Optional max target tokens. If not set, we default to 128
+    :param create_data_embs: Whether to create data embeddings for this run. Default
+        False
+    """
+
+    model: str = "google/flan-t5-base"
+    epochs: int = 3
+    learning_rate = 3e-4
+    accumulation_steps = 4
+    batch_size = 4
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/schema.py:72]:
```suggestion
    learning_rate: float = 3e-4
    accumulation_steps: int = 4
    batch_size: int = 4
```

File: dataquality/integrations/seq2seq/schema.py [hunk 5]
```diff
@@ -1 +1,94 @@
...
+class AutoTrainingConfig:
+    """Configuration for training a seq2seq model
+
+    :param model: The pretrained AutoModel from huggingface that will be used to
+        tokenize and train on the provided data. Default google/flan-t5-base
+    :param num_train_epochs: Optional num training epochs. If not set, we default to 3
+    :param max_input_tokens: Optional max input tokens. If not set, we default to 512
+    :param max_target_tokens: Optional max target tokens. If not set, we default to 128
+    :param create_data_embs: Whether to create data embeddings for this run. Default
+        False
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/schema.py:65]:
```suggestion
    :param create_data_embs: Whether to create data embeddings for this run. If set to
        None, data embeddings will be created only if a GPU is available
```

File: dataquality/integrations/seq2seq/s2s_trainer.py [hunk 6]
```diff
@@ -107,12 +98,12 @@ def get_trainer(
...
-    max_input_tokens = max_input_tokens or MAX_INPUT_LENGTH
-    max_target_tokens = max_target_tokens or MAX_TARGET_LENGTH
+    max_input_tokens = training_config.max_input_tokens
+    max_target_tokens = training_config.max_target_tokens
     tokenizer = AutoTokenizer.from_pretrained(
-        model_checkpoint, use_fast=True, model_max_length=max_input_tokens
+        training_config.model, use_fast=True, model_max_length=max_input_tokens
     )
-    model = T5ForConditionalGeneration.from_pretrained(model_checkpoint)
+    model = T5ForConditionalGeneration.from_pretrained(training_config.model)
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/s2s_trainer.py:106]: We'll have to find a way to load other models than T5s from HF at some point! Later later

File: dataquality/integrations/seq2seq/auto.py [hunk 7]
```diff
@@ -84,47 +81,25 @@ def auto(
...
-        set. If neither val nor test are available, the train data will be randomly
-        split 80/20 for use as evaluation data.
-        Can be one of
-        * Pandas dataframe
-        * Huggingface dataset
-        * Path to a local file
-        * Huggingface dataset hub path
-    :param test_data: Optional test data to use. The test data, if provided with val,
+    The validation data is what is
+    used for the evaluation dataset in huggingface, and what is used for early
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/auto.py:85]: We actually don't do early stopping in the training loop below. Created this ticket to maybe add it https://app.shortcut.com/galileo/story/8126 (not crucial in my opinion and hard to tune)

Comment (elboy3) [reply dataquality/integrations/seq2seq/auto.py:85]: true

File: dataquality/integrations/seq2seq/s2s_trainer.py (hunk 6 above)
Comment (elboy3) [reply dataquality/integrations/seq2seq/s2s_trainer.py:101]: oh yeah, definitely! good call

**Review by bogdan-galileo — APPROVED** [review]
Super clean! 🥇

File: dataquality/integrations/seq2seq/auto.py [hunk 8]
```diff
@@ -61,70 +65,43 @@ def _log_dataset_dict(dd: DatasetDict, input_col: str, target_col: str) -> None:
...
-        Can be one of
-        * Pandas dataframe
-        * Huggingface dataset
-        * Path to a local file
-        * Huggingface dataset hub path
-    :param num_train_epochs: Optional num training epochs. If not set, we default to 3
-    :param hf_model: The pretrained AutoModel from huggingface that will be used to
-        tokenize and train on the provided data. Default distilbert-base-uncased
+    One of DatasetConfig `hf_data`, `train_path`, or `train_data` should be provided.
+    If none of those is, a demo dataset will be loaded by Galileo for training.
```
Comment (dcaustin33) [inline dataquality/integrations/seq2seq/auto.py:82]: Cool! Do we do this demo dataset for other auto or just seq2seq

File: dataquality/integrations/seq2seq/s2s_trainer.py [hunk 9]
```diff
@@ -133,27 +118,22 @@ def get_trainer(
             ds_tokenized,
             shuffle=shuffle,
             collate_fn=data_collator,
-            batch_size=BATCH_SIZE,
+            batch_size=training_config.batch_size,
```
Comment (dcaustin33) [inline dataquality/integrations/seq2seq/s2s_trainer.py:121]: I wonder if there is some smart library for finding the optimal batch size for a model and a context length

**Review by dcaustin33 — APPROVED** [review]
Looks good to me! 🚀

File: dataquality/integrations/seq2seq/auto.py (hunk 8 above)
Comment (elboy3) [reply dataquality/integrations/seq2seq/auto.py:82]: yeah we do, it just picks a random one from DEMO_DATASETS

File: dataquality/integrations/seq2seq/auto.py [hunk 10]
```diff
@@ -61,70 +65,43 @@ def _log_dataset_dict(dd: DatasetDict, input_col: str, target_col: str) -> None:
...
+
+    The validation data is what is used for the evaluation dataset in huggingface.
```
Comment (setu4993) [inline dataquality/integrations/seq2seq/auto.py:84]: What does "evaluation dataset in huggingface" mean?

File: dataquality/integrations/seq2seq/auto.py [hunk 11]
```diff
@@ -61,70 +65,43 @@ def _log_dataset_dict(dd: DatasetDict, input_col: str, target_col: str) -> None:
...
+    If not provided, but test_data is, that will be used as the evaluation
+    set. If neither val nor test are available, the train data will be randomly
+    split 80/20 for use as evaluation data.
+
+    The test data, if provided with val,
+    will be used after training is complete, as the held-out set. If no validation
```
Comment (setu4993) [inline dataquality/integrations/seq2seq/auto.py:90]:
```suggestion
    will be used after training is complete, as the hold-out set. If no validation
```

File: dataquality/integrations/seq2seq/schema.py [hunk 12]
```diff
@@ -1 +1,95 @@
...
+    """Configuration for training a seq2seq model
+
+    :param model: The pretrained AutoModel from huggingface that will be used to
+        tokenize and train on the provided data. Default google/flan-t5-base
+    :param epochs: Optional num training epochs. If not set, we default to 3
+    :param learning_rate: Optional learning rate. If not set, we default to 3e-4
+    :param accumulation_steps: Optional accumulation steps. If not set, we default to 4
+    :param batch_size: Optional batch size. If not set, we default to 4
+    :param create_data_embs: Whether to create data embeddings for this run. If set to
+        None, data embeddings will be created only if a GPU is available
```
Comment (setu4993) [inline dataquality/integrations/seq2seq/schema.py:63]: Nice.

File: dataquality/integrations/seq2seq/schema.py [hunk 13]
```diff
@@ -1 +1,95 @@
...
+    """
+
+    model: str = "google/flan-t5-base"
+    epochs: int = 3
+    learning_rate: float = 3e-4
+    accumulation_steps: int = 4
+    batch_size: int = 4
+    create_data_embs: Optional[bool] = None
+    max_input_tokens: int = 512
+    max_target_tokens: int = 128
```
Comment (setu4993) [inline dataquality/integrations/seq2seq/schema.py:75]: Should this be output tokens instead of target tokens?

File: dataquality/integrations/seq2seq/schema.py [hunk 14]
```diff
@@ -1 +1,95 @@
...
+@dataclass
+class AutoGenerationConfig:
+    """Configuration for generating insights from a trained seq2seq model
+
+    We use the default values in HF GenerationConfig
+    See more about the parameters here:
+    https://huggingface.co/docs/transformers/v4.30.0/en/main_classes/text_generation#transformers.GenerationConfig
+
+    :param generation_splits: Optional list of splits to generate on. If not set, we
+        default to ["test"]
```
Comment (setu4993) [inline dataquality/integrations/seq2seq/schema.py:87]: Maybe worth documenting what happens if the test split is missing?

File: dataquality/integrations/seq2seq/s2s_trainer.py (hunk 9 above)
Comment (setu4993) [reply dataquality/integrations/seq2seq/s2s_trainer.py:121]: There's a couple that do that (IIRC `pytorch-lightning` had an extension to make it easy to find it) but it was always a bit weird to use with generative / s2s models because those tests don't try to generate using beam search, which blows up memory consumption.

File: dataquality/integrations/seq2seq/s2s_trainer.py (hunk 6 above)
Comment (setu4993) [reply dataquality/integrations/seq2seq/s2s_trainer.py:101]: Yeah, that time we'll also need different configs for the training since they'd likely need to have different batch sizes.

**Review by setu4993 — APPROVED** [review]
Looks great! Left a couple tiny notes which maybe worth addressing before merging.

File: dataquality/integrations/seq2seq/auto.py (hunk 10 above)
Comment (elboy3) [reply dataquality/integrations/seq2seq/auto.py:84]: whoops, that should be `in training`, fixing

File: dataquality/integrations/seq2seq/schema.py (hunk 13 above)
Comment (elboy3) [reply dataquality/integrations/seq2seq/schema.py:75]: we are using target everywhere

File: dataquality/integrations/seq2seq/schema.py (hunk 14 above)
Comment (elboy3) [reply dataquality/integrations/seq2seq/schema.py:87]: if missing it won't generate on test, do you think that's unclear?

Comment (setu4993) [reply dataquality/integrations/seq2seq/schema.py:87]: A bit. Would be nice to make explicit since we have some logic on the dataset side on how the validation set is created.
